server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java

```diff
@@ -134,8 +133,21 @@ private void handlePrimaryResult(final PrimaryResultT primaryResult) {
             markUnavailableShardsAsStale(replicaRequest, replicationGroup);
             performOnReplicas(replicaRequest, globalCheckpoint, maxSeqNoOfUpdatesOrDeletes, replicationGroup);
         }
-        successfulShards.incrementAndGet();  // mark primary as successful
-        decPendingAndFinishIfNeeded();
+        primaryResult.runPostReplicationActions(new ActionListener<>() {
+
+            @Override
+            public void onResponse(Void aVoid) {
+                updateCheckPoints(primary.routingEntry(), primary::localCheckpoint, primary::globalCheckpoint);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                logger.trace("[{}] op [{}] post replication actions failed for [{}]", primary.routingEntry().shardId(), opType, request);
+                // TODO: fail shard? This will otherwise have the local / global checkpoint info lagging, or possibly have replicas
```

Comment:
Yes, we should fail, AFAICS, the only reason we would go here is if the fsync fails and that should fail the shard asap?